File: 0480-sliding-window-median/0480-sliding-window-median.cpp
class Solution {
public:
    vector<double> medianSlidingWindow(vector<int>& nums, int k) {
        vector<double> res;
        vector<long long> ans;
        for(int i = 0 ; i < k; i++){
            ans.insert(lower_bound(ans.begin(), ans.end(), nums[i]), nums[i]);
        }
        if(k % 2==0){
            res.push_back( (double) (ans[k/2] + ans[k/2-1])/ 2.0);
        } else {
            res.push_back(ans[k/2]);
        }
        for(int i = k ; i < nums.size();i++){
            ans.erase(lower_bound(ans.begin(), ans.end(),nums[i-k]));
            ans.insert(lower_bound(ans.begin(), ans.end(), nums[i]), nums[i]);
            if(k % 2==0){
                res.push_back( (double) (ans[k/2] + ans[k/2-1])/ 2.0);
            } else {
                res.push_back(ans[k/2]);
            }            
        }
       return res;
    }
};

File: 0480-sliding-window-median/0480-sliding-window-median.java
class Solution {
    public double[] medianSlidingWindow(int[] nums, int k) {
        double[] res = new double[nums.length-k+1];
        int index = 0;
        PriorityQueue<Integer> left = new PriorityQueue<Integer>(Collections.reverseOrder());
        PriorityQueue<Integer> right = new PriorityQueue<Integer>();
        for(int i = 0; i < nums.length ; i++){
            left.add(nums[i]);
            right.add(left.poll());
            if(right.size() > left.size()) left.add(right.poll());
            if(right.size() + left.size() == k) {
                if(left.size() == right.size()){
                    res[index] = (double)((long)left.peek() + (long)right.peek()) /2;
                } else {
                    res[index] =  (double)left.peek();
                }
                if(!left.remove(nums[index])) right.remove(nums[index]);
                index++;
            }
            
        }
        return res;
    }
}